From b2ba87f471f8abd0066cae8c99014c7a52603e20 Mon Sep 17 00:00:00 2001
From: Wei Li <liwei391@huawei.com>
Date: Sun, 24 Apr 2022 11:29:56 +0800
Subject: [PATCH 0065/2065] [Backport] arm64: clear_page() shouldn't use DC ZVA
 when DCZID_EL0.DZP == 1

hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I53K4K

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=f0616abd4e67143b45b04b565839148458857347
Reference: https://gitee.com/openeuler/kernel/commit/ca467eff8b8fe66805c43b31d2a934312b605347

-------------------------------------------------

Currently, clear_page() uses DC ZVA instruction unconditionally.  But it
should make sure that DCZID_EL0.DZP, which indicates whether or not use
of DC ZVA instruction is prohibited, is zero when using the instruction.
Use STNP instead when DCZID_EL0.DZP == 1.

Fixes: f27bb139c387 ("arm64: Miscellaneous library functions")
Based-on-patch-by: Reiji Watanabe <reijiw@google.com>
Signed-off-by: Wei Li <liwei391@huawei.com>
Reviewed-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
---
 arch/arm64/kernel/cpufeature.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/arch/arm64/kernel/cpufeature.c b/arch/arm64/kernel/cpufeature.c
index 809dcac24e18..7d110ce5d7f4 100644
--- a/arch/arm64/kernel/cpufeature.c
+++ b/arch/arm64/kernel/cpufeature.c
@@ -1792,6 +1792,9 @@ static bool has_mor_nontemporal(const struct arm64_cpu_capabilities *entry)
 static bool can_clearpage_use_stnp(const struct arm64_cpu_capabilities *entry,
 				   int scope)
 {
+	if (read_sysreg(dczid_el0) & BIT(DCZID_DZP_SHIFT))
+		return true;
+
 	return use_clearpage_stnp && has_mor_nontemporal(entry);
 }
 
-- 
2.17.1

